sw/device/lib/dif/dif_spi_device.c

```diff
@@ -0,0 +1,695 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_spi_device.h"
+
+#include "spi_device_regs.h"  // Generated.
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/memory.h"
+
+const uint16_t kDifSpiDeviceSramLen = SPI_DEVICE_BUFFER_SIZE_BYTES;
+
+/**
+ * Computes the required value of the control register from a given
+ * configuration.
+ */
+static dif_spi_device_result_t build_control_word(
+    const dif_spi_device_config_t *config, uint32_t *device_config) {
+  *device_config = 0;
+  switch (config->clock_polarity) {
+    case kDifSpiDeviceEdgePositive:
+      *device_config |= 0x0 << SPI_DEVICE_CFG_CPOL;
+      break;
+    case kDifSpiDeviceEdgeNegative:
+      *device_config |= 0x1 << SPI_DEVICE_CFG_CPOL;
+      break;
+    default:
+      return kDifSpiDeviceResultBadArg;
+  }
+  switch (config->data_phase) {
+    case kDifSpiDeviceEdgePositive:
+      *device_config |= 0x1 << SPI_DEVICE_CFG_CPHA;
+      break;
+    case kDifSpiDeviceEdgeNegative:
+      *device_config |= 0x0 << SPI_DEVICE_CFG_CPHA;
+      break;
+    default:
+      return kDifSpiDeviceResultBadArg;
+  }
+  switch (config->tx_order) {
+    case kDifSpiDeviceBitOrderMsbToLsb:
+      *device_config |= 0x0 << SPI_DEVICE_CFG_TX_ORDER;
+      break;
+    case kDifSpiDeviceBitOrderLsbToMsb:
+      *device_config |= 0x1 << SPI_DEVICE_CFG_TX_ORDER;
+      break;
+    default:
+      return kDifSpiDeviceResultBadArg;
+  }
+  switch (config->rx_order) {
+    case kDifSpiDeviceBitOrderMsbToLsb:
+      *device_config |= 0x0 << SPI_DEVICE_CFG_RX_ORDER;
+      break;
+    case kDifSpiDeviceBitOrderLsbToMsb:
+      *device_config |= 0x1 << SPI_DEVICE_CFG_RX_ORDER;
+      break;
+    default:
+      return kDifSpiDeviceResultBadArg;
+  }
+
+  *device_config = bitfield_set_field32(
+      *device_config, (bitfield_field32_t){
```

Comment:
Is there a reason other then personal preference, why you always trying to avoid a temporary variable, in this case `bitfield_field32_t something`?

Reply:
No. Options structs are typically best passed as temporaries, since they're intended to be similar in spirit to python's kwargs.